source/win32/low_latency_iocp_context.cpp

```diff
@@ -0,0 +1,682 @@
+            std::terminate();
+        }
+    }
+}
+
+bool low_latency_iocp_context::try_allocate_io_state_for(io_operation* op) noexcept {
+    assert(is_running_on_io_thread());
+
+    if (ioFreeList_.empty()) {
+        return false;
+    }
+
+    assert(pendingIoQueue_.empty());
+
+    // An operation is already available
+    auto* state = ioFreeList_.pop_front();
+    state->parent = op;
+    state->completed = false;
+    state->operationCount = 0;
+    state->pendingCompletionNotifications = 0;
+    op->ioState = state;
+
+    return true;
+}
+
+void low_latency_iocp_context::schedule_when_io_state_available(io_operation* op) noexcept {
+    assert(is_running_on_io_thread());
+    assert(ioFreeList_.empty());
+    pendingIoQueue_.push_back(op);
+}
+
+void low_latency_iocp_context::release_io_state(vectored_io_state* state) noexcept {
+    assert(is_running_on_io_thread());
+
+    if (state->pendingCompletionNotifications == 0) {
+        // Can be freed immediately.
+
+        if (pendingIoQueue_.empty()) {
+            ioFreeList_.push_front(state);
+        } else {
+            assert(ioFreeList_.empty());
+
+            // Another operation was waiting for an I/O state.
+            // Give the I/O state directly to the operation instead
+            // of adding it to the free-list. This prevents some other
+            // operation from skipping the queue and acquiring it
+            // before this I/O operation can run.
+            auto* op = static_cast<io_operation*>(pendingIoQueue_.pop_front());
+
+            state->parent = op;
+            state->completed = false;
+            state->operationCount = 0;
+            state->pendingCompletionNotifications = 0;
+            op->ioState = state;
+
+            schedule_local(op);
+        }
+    } else {
+        // Mark it to be freed once all of its I/O completion
+        // notifications have been received.
+        state->parent = nullptr;
+    }
+}
+
+void low_latency_iocp_context::schedule_poll_io(io_operation* op) noexcept {
+    assert(is_running_on_io_thread());
+    assert(op != nullptr);
+    assert(op->ioState != nullptr);
+
+    if (op->ioState->pendingCompletionNotifications > 0) {
+        pollQueue_.push_back(op);
+    } else {
+        // No need to poll, schedule straight to the front of the ready-to-run queue.
+        readyQueue_.push_front(op);
+    }
+}
+
+void low_latency_iocp_context::associate_file_handle(handle_t fileHandle) {
+    const HANDLE result = ::CreateIoCompletionPort(
+        fileHandle,
+        iocp_.get(),
+        0, 0);
+    if (result == nullptr) {
+        DWORD errorCode = ::GetLastError();
+        throw std::system_error{
+            static_cast<int>(errorCode), std::system_category(),
+            "CreateIoCompletionPort"};
+    }
+
+    const BOOL ok = ::SetFileCompletionNotificationModes(
+        fileHandle,
+        FILE_SKIP_COMPLETION_PORT_ON_SUCCESS |
+        FILE_SKIP_SET_EVENT_ON_HANDLE);
+    if (!ok) {
```

Comment:
You can't do this unfortunately. Whether skipping completion port on success is available or not depends on the NIC driver. Some NIC drivers don't implement it. So you really do need to fallback if it fails.

Reply:
This is specifically for named pipes and files, which afaict should always support skipping completion port...?

I am aware of the need to query the capability of this for sockets, but have not yet added socket support.

When I do I plan to add a new `associate_socket()` function that does the capability a check as I did in [cppcoro](https://github.com/lewissbaker/cppcoro/blob/69e575480afe97fd342a44f580ba8ff3b4a9a58a/lib/socket.cpp#L146).